Syscall forwarding refactored from #21

Thaodan opened the pull request:
Originally based on work by Frajo Haider but completely reworked
to use the existing interface system.

Contains #32, obsoletes #21.

Thaodan committed:
[sb2] Allow redirecting syscalls (not all are supported yet). JB#59837

Originally based on work by Frajo Haider but completely reworked
to use the existing interface system.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Allow redirecting syscalls (not all are supported yet). JB#59837

Originally based on work by Frajo Haider but completely reworked
to use the existing interface system.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Avoid errors in out of tree builds for files from $(OBJDIR)

Previously in some instance the assumption was $(SRCDIR) = $(OBJDIR).
If $(OBJDIR) isn't equal to $(SRCDIR) these files were not found anymore.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[utils] Use va_args in open_nomap_nolog to match wrappers. JB#61866

Resolves error when using a compiler with stricter settings such as
when using LTO.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[preload] Adapt glibc in open ignore 'mode' with O_TMPFILE in flags. JB#61866

Both open and openat load their last argument 'mode' lazily, using
va_arg() only if O_CREAT is found in oflag. This is wrong, mode is also
necessary if O_TMPFILE is in oflag.

Adapt __OPEN_NEEDS_MODE(oflag) present since glibc 2.22 to solve the
issue.

Read more down below:
https://sourceware.org/git/?p=glibc.git;a=commit;h=65f6f938cd562a614a68e15d0581a34b177ec29d

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Don't run make twice

Once in autogen.sh and once in %make

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[preload] Fix generated dependency order

Previously preload/exported.h depended on preload/wrappers.c even thou
both where generated from interface.master by gen-interface.pl.

Refactor so that all generated preloader files depend on the interface
and the gen-interface script.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[llbuild] Refactor import_dir to match other Makefiles

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Remove multilib for x86_64 and use autotools variables. JB#61868

The builtin multilib support as unused. It's cleaner to built
libsb2 for 32bit targets in a 32bit worker.
The side effect of these changes is that make file pickup
the correct directories passed to configure in %configure.

Also now we can use regular build macros.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Use autotools builtin system to rerun configure instead of our own

By detecting if configure.ac has changed and thus configure will change
in the following step we can run autogen.sh and config.status to rerun
configure with the previously set built options.

The change removes the need for .configure file, correctly detects
if configures sources changed and doesn't run configure again without
all the previously passed options.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Automatically detect out of tree builds

In case the root Makefile isn't in PWD assume that the Makefiles
directory is the source directory.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
packaging: Cleanup spec file

Remove obsolete and unused things.
Own all directories correctly (this will be an error in later RPM versions).

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
sb2: Fix errors found shellcheck SC2166, SC2181, SC2048 and SC2012

Remove useless checks like:
```
command
if [ $# -eq 0 ] ; then
    error
fi
```

Or replace obsolescent -a and -o with `[ ] && [ ] ` and `[ ] || [ ]`.

https://github.com/koalaman/shellcheck/wiki/SC2166
https://github.com/koalaman/shellcheck/wiki/SC2181

Thaodan committed:
[sb2] Check if default mode for target exists, error out earlier

Doing so exits with a clearer error message then errors about sb2d
failing to start later such as these:

Runtime-Error loading /tmp/sb2-bidar-20240329-145433.AwoNRb/lua_scripts/init.lua (...dar-20240329-145433.AwoNRb/lua_scripts/argvenvp_misc.lua:28: bad argument #1 to 'files' (/tmp/sb2-bidar-20240329-145433.AwoNRb/wrappers.sdk-build: No such file or directory))
sb2: Error: startup of sb2d failed.

Tested-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Take directory variables from configure. JB#61868

To make it easier to detect where Scratchbox2 helper scripts
and binaries are found on the host and partially on the
tooling/target.
This is especially important for helper binaries such as sb2dctl.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[preloader] Add gates for wait3 and wait4

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[preloader] Add gates for wait3 and wait4

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
sb2: Remove trailing slash

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Take version from autoconf to avoid unintended rebuilds

The version used in the Makefile and version header was extracted
using git with a phony target which resulted in unintended rebuilds.

Using the autoconf script avoids these as now the version only updates
when the autoconf script changes.

Further the version originally set in the autoconf was also out of
date. This change fixes both of these issues.
The script in build-aux was taken from Gnulib.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Clean correctly without breaking builds and keeping the tree dirty

Previously config.status was also clean as part of the clean making
rebuilds impossible without calling configure again.
Clean isn't supposed to clean the generated build- or
distribution-system files.

Further somefiles that should be clean where not cleaned at all.
This change also aligns the name of the clean target which removes
the distribution to the one configure mentions when it fails.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Make sure config.h is updated and cleaned

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Avoid rebuilds when configure.ac changes but not it's output

When configure.ac or config.mak.in change but not their output their
will be repeated rebuilds as the timestamp their output files such as
config.mak stay older than their dependency.

The timestamp file's timestamp will always update after the targets
dependency changed.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

direc85 committed:
sb2: Move logic to functions, session checks to parser, shorten code

Co-authored-by: Matti Viljanen <matti.viljanen@jolla.com>
Co-authored-by: Björn Bidar <bjorn.bidar@jolla.com>
Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
packaging: Also include x86_64 in ExclusiveArch

In case not available.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
Add EditorConfig and dir-locals

Sets basic indentation across editors and reviews on Github.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
preload: Remove mentions of ldexportlist or replace with export.map

ldexportlist was removed earlier for export.map. Replace with
export.map where not done already. Remove obsolete references such as
the target in preload/Makefile.

Fixes: 20180ac890cd54ccc1ee1b450a660681e87ce1bd
Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sblib] Fix potential segfaults in sb_log

The char arrays in sblog_vprintf_line_to_logfile where not initialized
which can cause the the call to vsnprintf can fail.
See for example the backtrace below:

   from ~/dev/vms/jolla/targets/SailfishOS-devel-i486/lib/libc.so.6
    level=9, format=0xf7f6c488 "%s gate: calling mapped syscall(%s) to %s",
    ap=0xffcd9f10 "\304\035\367\367\005")
    at ~/dev/sailfish/jolla/pkgs/pj_oss/scratchbox2/sblib/sb_log.c:318
    level=9, format=0xf7f6c488 "%s gate: calling mapped syscall(%s) to %s")
    at ~/dev/sailfish/jolla/pkgs/pj_oss/scratchbox2/sblib/sb_log.c:423
    realfnname=0xf7f71988 <__func__.292> "syscall", number=5, ap=0xffcd9fc4 "\034\240\004\b")
    at preload/syscallgate.c:91

Thaodan committed:
[sblog] Add format attribute to detect type conversion errors at compile time

Thaodan committed:
[sb2] Make format string specifiers match the target architecture

Also remove some extra white space at end or in empty lines.

piggz commented:
This appears to fix the issues building qt6, with file permission errors. See https://build.sailfishos.org/project/monitor/home:piggz:qt6sb2?arch_aarch64=1&defaults=0&repo_sailfish_latest_aarch64=1&succeeded=1  i had to fork and make some build fixes, minor printf formatting placeholders.

vige commented:
You should rebase the PR on top of current master. It would still be a huge PR but at least a bit more comprehensible.

piggz commented on execs/exec_policy_ruletree.c:
```diff
@@ -52,10 +54,10 @@ const char *exec_policy_get_string(exec_policy_handle_t eph,
 		if (str) {
 			SB_LOG(SB_LOGLEVEL_NOISE,
-				"%s: %s='%s'", __func__, s_name, str, fldoffs);
+				"%s: %s='%s' o=%lu", __func__, s_name, str, fldoffs);
```
Should be %u for size_t

piggz replied:
Discussed with @mlehtima , %zu is probably better?

piggz commented on execs/exec_policy_ruletree.c:
```diff
@@ -52,10 +54,10 @@ const char *exec_policy_get_string(exec_policy_handle_t eph,
 		} else {
 			SB_LOG(SB_LOGLEVEL_NOISE,
-				"%s: No %s", __func__, s_name, fldoffs);
+				"%s: No %s o=%lu", __func__, s_name, fldoffs);
```
Should be %u

piggz commented on execs/exec_policy_ruletree.c:
```diff
@@ -73,7 +75,7 @@ int exec_policy_get_boolean(exec_policy_handle_t eph,
 		uip = ruletree_get_pointer_to_boolean(offs);
                 SB_LOG(SB_LOGLEVEL_NOISE,
-                        "%s: @%u = *%p = %d",
+                        "%s: @%u = *%p = %d o=%lu",
```
Should be %u

piggz commented on execs/exec_policy_ruletree.c:
```diff
@@ -91,7 +93,7 @@ int exec_policy_get_uint32(exec_policy_handle_t eph,
 		uip = ruletree_get_pointer_to_uint32(offs);
 		SB_LOG(SB_LOGLEVEL_NOISE,
-			"%s: @%u = *%p = %u",
+			"%s: @%u = *%p = %u o=%lu",
```
Should be %u

piggz commented on execs/exec_policy_ruletree.c:
```diff
@@ -107,9 +109,8 @@ ruletree_object_offset_t exec_policy_get_rules(exec_policy_handle_t eph,
 			eph.exec_policy_offset, r_name);
                 SB_LOG(SB_LOGLEVEL_NOISE,
-                        "%s: @%u", __func__, offs, fldoffs);
+                        "%s: @%u o=%lu", __func__, offs, fldoffs);
```
Should be %u

piggz commented on execs/exec_policy_selection.c:
```diff
@@ -87,7 +91,7 @@ static int test_path_match(const char *full_path, size_t full_path_len,
 
 	SB_LOG(SB_LOGLEVEL_NOISE,
-		"%s: '%s' (%u), '%s' (%u) => %d (%s)",
+		"%s: '%s' (%lu), '%s' (%u) => %d (%s)",
```
Should be %u

piggz commented on pathmapping/paths_ruletree_mapping.c:
```diff
@@ -109,7 +109,7 @@ static int ruletree_test_path_match(const char *full_path, size_t full_path_len,
 
 	SB_LOG(SB_LOGLEVEL_NOISE,
-		"ruletree_test_path_match '%s' (%u), '%s' (%u) => %d (%s)",
+		"ruletree_test_path_match '%s' (%lu), '%s' (%u) => %d (%s)",
```
Should be %u

piggz commented on rule_tree/rule_tree.c:
```diff
@@ -99,16 +99,16 @@ ruletree_object_offset_t append_struct_to_ruletree_file(void *ptr, size_t size,
 		if (write(ruletree_ctx.rtree_ruletree_fd, ptr, size) < (int)size) {
 			SB_LOG(SB_LOGLEVEL_ERROR,
-				"Failed to append a struct (%d bytes) to the rule tree", size);
+				"Failed to append a struct (%lu bytes) to the rule tree", size);
```
Should be %u

piggz commented on rule_tree/rule_tree.c:
```diff
@@ -370,13 +370,13 @@ ruletree_object_offset_t ruletree_objectlist_create_list(uint32_t size)
 	if ((wr_result == -1) || ((size_t)wr_result < list_size_in_bytes)) {
 		SB_LOG(SB_LOGLEVEL_ERROR,
-			"Failed to append a list (%d items, %d bytes) to the rule tree", 
+			"Failed to append a list (%d items, %lu bytes) to the rule tree",
```
Should be %u

piggz commented on rule_tree/rule_tree.c:
```diff
@@ -656,7 +656,7 @@ int ruletree_find_inodestat(
 
 	SB_LOG(SB_LOGLEVEL_NOISE,
-		"ruletree_find_inodestat (dev=%lld,ino=%lld,key=%llX)",
+		"ruletree_find_inodestat (dev=%lld,ino=%lld,key=%lu)",
```
Should be %llu

piggz commented on rule_tree/rule_tree.c:
```diff
@@ -698,7 +698,7 @@ ruletree_object_offset_t ruletree_set_inodestat(
 {
 	SB_LOG(SB_LOGLEVEL_NOISE,
-		"ruletree_set_inodestat (dev=%lld,ino=%lld,key=%llX))",
+		"ruletree_set_inodestat (dev=%lld,ino=%lld,key=%lu))",
```
Should be %llu

Thaodan committed:
[sb2] Wrap rule typedef into its own PRI macro, use PRI macros consistently

Thaodan committed:
[preloader] Don't link lua into libsb2

When Lua was unbundled, i.e. removed from libsb2, liblua was
accidentally linked to lisb2.
That added liblua as a dependency to targets even thou it wasn't used in them.

Fixes: c967c03c5be743eb23587f00f0dfe9a6c06ece7a

Thaodan committed:
[sb2] Redo version handling, determine soname at runtime

Before the version was defined in an extra header but also defined in
config.h which includes the version anyway as it's put in there through
the autoconf script.
Doing so avoid race conditions during parallel builds.

The version header also defined the sover. The sover was defined
in multiple places before, both in the build as well as in the shell.
This inconsistency meant that in some places it said the sover was one
in others it was two.

Now the sover is set in configure.ac once for both C code aswell as the
rest.

The actual soname is set once during build for itself and target
initialization.
After that it's determined at runtime through the ELF interpreter.

Thaodan committed:
[sb2] Define the preloader per policy context

Enable the preloader to be different depending on the policy context,
i.e. for backwards compatibility and for testing.
